Add framework-side hitTestBehavior support for Semantics widget and apply to ModalRoute

flutter-zl opened the pull request:
Fix premature dialog dismissal on Flutter Web when semantics are enabled by correctly propagating hitTestBehavior through the semantics pipeline and declaring modal routes as opaque to pointer events.

Before change
https://dialog-dismiss-before.web.app/

Click on the "Show Dialog" button.
Click anywhere inside the dialog that is not a form field.
Observe the dialog being dismissed.

After change
https://dialog-dimiss-after.web.app/

Click on the "Show Dialog" button.
Click anywhere inside the dialog that is not a form field.
Observe the dialog not dismissed.

Fixes: #149001
Engine work: #176974

flutter-zl committed:
Add framework-side hitTestBehavior support for Semantics widget and apply to ModalRoute

flutter-zl committed:
Fix dialog dismissal bug by adding _hitTestBehavior to SemanticsConfiguration.copy() method

flutter-zl committed:
Add tests for SemanticsConfiguration.copy() hitTestBehavior preservation

flutter-zl committed:
Address gemini-code-assist review: add hitTestBehavior to equality/hashCode/absorb/merging

flutter-zl committed:
Fix: add missing hitTestBehavior argument in semantics tests to fix analyzer warnings.

flutter-zl committed:
Fix(format): remove trailing space in semantics.dart

chunhtai commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4127,6 +4130,7 @@ sealed class _SemanticsBase extends SingleChildRenderObjectWidget {
            role: role,
            controlsNodes: controlsNodes,
            validationResult: validationResult,
+           hitTestBehavior: hitTestBehavior,
```
we may need to force node that sets this to not merge up

For example
```dart
Column(
  children: [
     Text('child1'),
     Semantics(hitTestBehavior: .opaque, child: Text('child2')),
     Text('child3'),

  ]
)
```

If you print the semantics tree you will see the entire column will form a node, and the opaque will merge up to apply to the entire node

There are two way to fix this
1. force Semantics widget that sets hitTestBehavior to also have container:true
2. implement SemanticsConfiguration.isCompatible to reject the merge in this case

I think going with (2) makes more sense

flutter-zl committed:
Add hitTestBehavior compatibility checks to prevent unwanted merging

flutter-zl committed:
Fix dropdown_test.dart semantics expectations for hitTestBehavior changes

The modal barrier and dropdown menu now correctly appear as siblings
in the semantics tree instead of incorrectly merging, due to the
hitTestBehavior compatibility checks.

flutter-zl committed:
Format dropdown_test.dart

chunhtai reviewed (commented):
Can you add a test that verify the modal route has this new property?

flutter-zl committed:
add new test case and remove redundant hitTestBehavior from RawDialogRoute.buildPage

flutter-zl commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -2616,6 +2617,7 @@ class RawDialogRoute<T> extends PopupRoute<T> {
     return Semantics(
       scopesRoute: true,
       explicitChildNodes: true,
+      hitTestBehavior: ui.SemanticsHitTestBehavior.opaque,
```
Good catch. I updated code accordingly.

flutter-zl committed:
Fix linting issue: use if-null operator for nullable bool

flutter-zl committed:
fix test case issue

flutter-zl committed:
removed the trailing spaces

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -3671,6 +3693,9 @@ class SemanticsNode with DiagnosticableTreeMixin {
           inputType = node._inputType;
         }
+        if (node._hitTestBehavior != ui.SemanticsHitTestBehavior.defer) {
+          hitTestBehavior = node._hitTestBehavior;
```
This merging code is called when MergeSemantics widget is used, this means all semantics nodes under the tree are forced to form one semantics node. so compatibility check doesn't play a role here. FYI, compatibility check will affect whether SemanticsConfiguration.absorb is called

chunhtai replied:
for example this code will trigger this part of merge

```dart
MergeSemantics(
  child: Semantics(
     hitTestBehavior: .opaque,
     child: Semantics(
       hitTestBehavior: .transparent,
       child: SizedBox(width: 10, height: 10),
     ),
  ),
);
```

chunhtai reviewed (commented):
just one last concern

chunhtai reviewed (approved):
LGTM

chingjun commented:
Reason for revert: Broke internal tests.
